tests/subsys/bluetooth/fast_pair/storage/factory_reset/src/main.c

```diff
@@ -0,0 +1,424 @@
+	zassert_ok(err, "Unexpected error in factory reset");
+}
+
+static void test_normal_reset(void)
+{
+	if (test_stage != RESET_TEST_STAGE_NORMAL) {
+		ztest_test_pass();
+	}
+
+	store_data_and_factory_reset();
+
+	storage_validate();
+}
+
+static void test_normal_reset_flow_2(void)
+{
+	if (test_stage != RESET_TEST_STAGE_NORMAL) {
+		ztest_test_pass();
+	}
+
+	store_data_and_factory_reset();
+
+	storage_validate_flow_2();
+}
+
+static void test_interrupted_reset(void)
+{
+	int err;
+
+	switch (test_stage) {
+	case RESET_TEST_STAGE_NORMAL:
+		storage_reset_reboot = true;
+
+		test_stage = RESET_TEST_STAGE_INTERRUPT;
+		err = settings_save_one(SETTINGS_RESET_TEST_STAGE_FULL_NAME, &test_stage,
+					sizeof(test_stage));
+		zassert_ok(err, "Unexpected error in test settings");
+
+		store_data_and_factory_reset();
+
+		/* The recent instruction should trigger reboot so this code should never be
+		 * entered.
+		 */
+		ztest_test_fail();
+
+		break;
+
+	case RESET_TEST_STAGE_INTERRUPT:
+		storage_validate();
+
+		break;
+
+	case RESET_TEST_STAGE_INTERRUPT_FLOW_2:
+		ztest_test_pass();
+
+		break;
+
+	default:
+		ztest_test_fail();
+		break;
+	}
+}
+
+static void test_interrupted_reset_flow_2(void)
+{
+	int err;
+
+	switch (test_stage) {
+	case RESET_TEST_STAGE_NORMAL:
+		/* This shouldn't happen as the test_interrupted_reset test should change
+		 * the test_stage variable.
+		 */
+		ztest_test_fail();
+
+		break;
+
+	case RESET_TEST_STAGE_INTERRUPT:
+		storage_reset_reboot = true;
+
+		test_stage = RESET_TEST_STAGE_INTERRUPT_FLOW_2;
+		err = settings_save_one(SETTINGS_RESET_TEST_STAGE_FULL_NAME, &test_stage,
+					sizeof(test_stage));
+		zassert_ok(err, "Unexpected error in test settings");
+
+		store_data_and_factory_reset();
+
+		/* The recent instruction should trigger reboot so this code should never be
+		 * entered.
+		 */
+		ztest_test_fail();
+
+		break;
+
+	case RESET_TEST_STAGE_INTERRUPT_FLOW_2:
+		storage_validate_flow_2();
+
+		break;
+
+	default:
+		ztest_test_fail();
+		break;
+	}
+}
+
+void test_main(void)
+{
+	int err;
+
+	err = settings_subsys_init();
+	zassert_ok(err, "Error while initializing settings subsys");
+
+	/* The test_interrupted_reset and the test_interrupted_reset_flow_2 tests must be placed in
+	 * the suite in this exact places, because of dependencies during handling the test_stage
+	 * variable.
+	 */
+	ztest_test_suite(fast_pair_storage_factory_reset_tests,
+			 ztest_unit_test(test_section_placement),
+			 ztest_unit_test_setup_teardown(test_store, setup_fn, teardown_fn),
+			 ztest_unit_test_setup_teardown(test_store_flow_2, setup_fn, teardown_fn),
+			 ztest_unit_test_setup_teardown(test_normal_reset, setup_fn, teardown_fn),
+			 ztest_unit_test_setup_teardown(test_normal_reset_flow_2, setup_fn,
+							teardown_fn),
+			 ztest_unit_test_setup_teardown(test_interrupted_reset, setup_fn,
+							teardown_fn),
+			 ztest_unit_test_setup_teardown(test_interrupted_reset_flow_2, setup_fn,
+							teardown_fn)
+			 );
```

Comment:
Is it possible to define multiple test suites and run each one of them according to the ```test_stage``` value? This way, we could get rid of some of the boilerplate code like this one:

```
if (test_stage != RESET_TEST_STAGE_NORMAL) {
	ztest_test_pass();
}
```

Another benefit of this approach would be clean test logs which currently repeat (```test_store``` pass is repeated in each stage).